### ISSUE TITLE
Feature Request: Throw-by-default Error Handling

### ISSUE THREAD

**Issue by NfNitLoop** [nushell/nushell#10633]
### Related problem

I use Nushell's scripting language mostly to handle calling executables and parsing their output, or performing some set of bulk operations. But as I build more complicated scripts with Nushell, its default handling of errors from executables is starting to feel like a big footgun.

Many Nushell built-ins already handle errors like I'd expect, something like "throw" in other languages. Execution is stopped, and the user gets a detailed error:

```
> open asdfasdf | split row " " | print; print done
Error: nu::shell::directory_not_found

  × Directory not found
   ╭─[entry #63:1:1]
 1 │ open asdfasdf | split row " " | print; print done
   ·      ────┬───
   ·          ╰── directory not found
   ╰────
```

But not everything is a built-in.  If I need to reach for an executable to perform some function, the behavior changes:

```
cat asdfasdf | split row " " | print; print done
cat: asdfasdf: No such file or directory
╭───┬──╮
│ 0 │  │
╰───┴──╯

done
```

Now the error is swallowed. `cat` is nice enough to print an error to stderr, but it didn't stop execution. Nushell confidently returned the wrong value to me, and carried on with its execution.

It might not seem so bad in this toy example, where I can see the error message on screen. But other cases might have returned enough results to push the error message off of the screen.

If the usage of `cat` is buried in N levels of function calls, the distance between the place where the real error occurred (no such file) and where Nushell might eventually error (say, referencing a column/value that should exist) may be large, and difficult to debug.

### Describe the solution you'd like

1) Non-zero exit codes should "throw" (`error make`) everywhere by default. 

I'd expect `cat asdfasdf` to fail the same way that this fails today:

```
> do -c { cat asdfasdf }
cat: asdfasdf: No such file or directory
Error: nu::shell::external_command

  × External command failed
   ╭─[entry #71:1:1]
 1 │ do -c { cat asdfasdf }
   ·         ─┬─
   ·          ╰── External command failed
   ╰────
  help:
```

2) The user should be able to opt out of this behavior if they want. (`do -i/-p`, `try/catch`, or something.) But this behavior should be *local* and not inherited.

Ex: As an author of a nushell library, I should be able to write a function like:

```nushell
def doThing [] {
   print "Starting thing…"
   extern-tool foo
   extern-tool bar --baz  # Shouldn't be run if the previous line failed. Might do bad things in that case.
   print "Success!"
}
```

… and rely on the "throw by default" behavior of NuShell to exit my function with an error showing *that* extern-tool failed and *which call* failed, with no extra effort from me.

And an external user shouldn't be able to disable the default error handling which my function relies on.  I call this out because users *can* seem to do something similar with `do -c` and subexpressions currently:

```
~> def ex1 [] {
::: print start
::: cat asdfasdf
::: print done
::: }
~> ex1
start
cat: asdfasdf: No such file or directory
~> do -c { ex1 }
start
cat: asdfasdf: No such file or directory
Error: nu::shell::external_command

  × External command failed
   ╭─[entry #27:2:1]
 2 │ print start
 3 │ cat asdfasdf
   · ─┬─
   ·  ╰── External command failed
 4 │ print done
   ╰────
  help:

~> (ex1)
start
cat: asdfasdf: No such file or directory
done
```

3) There could be well-defined exceptions to throw-by-default.

For example, if you pipe directly to `| complete`, you explicitly have stderr and the exit code. The assumed intention is that you're going to check those. So you could skip the `error make` in this case. (If implementing an exception in pipes is a pain, maybe make a command like `exec` that captures the outputs?  ex: `capture cat asdfasdf`)

Simple commands run directly in the REPL (i.e. not inside of a function or block, or chain of commands) could forego `error make` under the assumption that they'll report their own errors. (ex: the [git log](https://www.jntrnr.com/exit-codes/) example.)

### Describe alternatives you've considered

Just wrap your calls to executables in `do -c { … }`
----------------

"Just don't shoot yourself with the foot-gun every time you and anyone else uses it."  😬 

But also: It's not always easy to know what is an executable vs. what's a built-in.  You might think that `git` is an executable, so `git branch` needs to be wrapped in a `do -c { … }`, but that's actually unnecessary in *my* shell because I have a custom `git branch` command.

IMO shell languages are nice when they blur the lines between built-in and binary commands. I'd love it if errors just operated the same way in both cases.

### Additional context and details

I originally had much of this conversation in #10624 but I'm moving the discussion here to let that ticket focus on the existing bug, and this issue can be the request for a change in the fundamental behavior.

**Comment by amtoine** [comment]
to be sure i understand, you propose something similar to what `do -ic { ... }` gives?
```nushell
> cat asdfasdf
Error: nu::shell::external_command

  × External command failed
   ╭─[entry #3:1:1]
 1 │ cat asdfasdf
   · ─┬─
   ·  ╰── External command failed
   ╰────
  help: cat: asdfasdf: No such file or
        directory
```

**Comment by NfNitLoop** [comment]
@amtoine Exactly. I think the important parts are that it should be the default behavior, not opt-in. (maybe with a few well-documented exceptions. see # 3.) Errors from binaries should be as nice as they are from nu functions.

And piping the output of commands shouldn't stop those errors from "bubbling up".

**Comment by amtoine** [comment]
sounds pretty sensible to me :+1:

**Comment by futile** [comment]
I just stumbled into the evil behavior around `do -c` and subexpressions mentioned here, and I would agree that it's currently really easy to shoot yourself in the foot.

Example:
```nushell
#!/usr/bin/env nu

def main [] {
  do -c {
    print "starting"
    (nu -c 'exit 1')
    print "I should not be printed #1"    
    do -c {
      (nu
        -c 'exit 1') 
    }
    print "I should not be printed #2"
    nu -c 'exit 1'
    print "I should not be printed #3"

  }
}
```

When run:
```console
❯ ./foo.nu
starting
I should not be printed #1
I should not be printed #2
Error: nu::shell::external_command

  × External command failed
    ╭─[/home/felix/gits/nocode/foo.nu:12:1]
 12 │     print "I should not be printed #2"
 13 │     nu -c 'exit 1'
    ·     ─┬
    ·      ╰── External command failed
 14 │     print "I should not be printed #3"
    ╰────
  help:
```

This still prints `#1` and `#2`, even though an exit code != 0 occurred before.

And another example, which is also really unintuitive:
```nushell
#!/usr/bin/env nu

def main [] {
  (do -c {
    nu -c 'exit 1'
    print "I should not be printed" 
  })
}
```
Runs as:
```console
❯ ./foo.nu
I should not be printed
```

While removing the `print` makes it throw the error:
```nushell
#!/usr/bin/env nu

def main [] {
  (do -c {
    nu -c 'exit 1'
  })
}
```
Runs as:
```console
❯ ./foo.nu
Error: nu::shell::external_command

  × External command failed
   ╭─[/home/felix/gits/nocode/foo.nu:4:1]
 4 │   (do -c {
 5 │     nu -c 'exit 1'
   ·     ─┬
   ·      ╰── External command failed
 6 │   })
   ╰────
  help:
```

**Comment by gergelyk** [comment]
I totally agree with @NfNitLoop. Wrapping each external command into `do -c { ... }` sounds way too verbose and error prone. Even bash allows for breaking pipelines on an error by using `set -o pipefail`. Also note that [elvish](https://elv.sh/) implements this behavior by default.

Additionally, from the [documentation](https://www.nushell.sh/commands/docs/do.html) it isn't clear what `-i` and `-c` options do exactly. They sound like complementary options. @amtoine, could you explain why this makes sense to apply both of them?

**Comment by amtoine** [comment]
@gergelyk 
this is a part of Nushell i'm not very familiar with unfortunately... :cry:

**Comment by amtoine** [comment]
i can see that the errors are not the same with and without `do -c`
```nushell
> do --capture-errors { nu --commands 'exit 1' } | myscarycommand
Error: nu::shell::external_command

  × External command failed
   ╭─[entry #4:1:1]
 1 │ do --capture-errors { nu --commands 'exit 1' } | myscarycommand
   ·                       ─┬
   ·                        ╰── External command failed
   ╰────
  help:
```
```nushell
> nu --commands 'exit 1' | myscarycommand
Error: nu::shell::external_command

  × External command failed
   ╭─[entry #5:1:1]
 1 │ nu --commands 'exit 1' | myscarycommand
   ·                          ───────┬──────
   ·                                 ╰── executable was not found
   ╰────
  help: No such file or directory (os error 2)
```

**Comment by amtoine** [comment]
and `-i` discards the error as expected
```nushell
> do --ignore-errors { nu --commands 'exit 1' } | myscarycommand
Error: nu::shell::external_command

  × External command failed
   ╭─[entry #6:1:1]
 1 │ do --ignore-errors { nu --commands 'exit 1' } | myscarycommand
   ·                                                 ───────┬──────
   ·                                                        ╰── executable was not found
   ╰────
  help: No such file or directory (os error 2)
```

**Comment by gergelyk** [comment]
No worries @amtoine. I was referring to your previous [post](https://github.com/nushell/nushell/issues/10633#issuecomment-1751676741) where you use `-i` and `-c` together. I thought that you may know it under the hood. 

It seems that `do` with no options does the same as when we add `-i`:

```nu
~ » do { nu --commands 'exit 1' } | myscarycommand
Error: nu::shell::external_command

  × External command failed
   ╭─[entry #20:1:1]
 1 │ do { nu --commands 'exit 1' } | myscarycommand
   ·                                 ───────┬──────
   ·                                        ╰── executable was not found
   ╰────
  help: No such file or directory (os error 2)
```
That's why I'm confused.

**Comment by amtoine** [comment]
i'm confused as well :laughing:

**Comment by tv42** [comment]
Related to / duplicate of #6617

**Comment by abusch** [comment]
I'm not saying there aren't improvements that are needed for error handling, but I have issues with the 2 examples in the description and your expectations. Nushell here behaves exactly like `bash`: when you have a pipeline of commands, it's the last command in the pipeline that determines the "success" or "error" of the whole pipelines. You can't prevent the following commands in the pipeline to run if the first one fails, because the first one could fail _at any time_, and until (or if) it does, you need to start streaming the output to the other commands down the pipeline.

See this example in bash:

```bash
bash-5.2$ cat /foo
cat: /foo: No such file or directory
bash-5.2$ echo $?
1
bash-5.2$ cat /foo | wc -l
cat: /foo: No such file or directory
       0
bash-5.2$ echo $?
0
bash-5.2$ (cat /foo | wc -l) && echo "hello"
cat: /foo: No such file or directory
       0
hello
bash-5.2$
```

**Comment by abusch** [comment]
Just to be clear, the 2 examples at the top behave exactly like `bash` does by default (unless you explicitly use `set -o pipefail`). So I guess what `nushell` is lacking is some equivalent to `set -o pipefail` and `set -e` (potentially turned on by default).

Also note that you can't prevent the subsequent commands in a pipeline to run when the first one fails (even with `set -o pipefail`), because it could fail at any time and until then you need to start streaming the output to the rest of the commands.

In the example below, `cat` fails, and its error code is properly reported, but `wc -l` still gets executed.
```bash
bash-5.2$ set -o pipefail
bash-5.2$ cat /foo | wc -l
cat: /foo: No such file or directory
       0
bash-5.2$ echo $?
1
```

**Comment by abusch** [comment]
The case of `(external)` is a bit insidious indeed. Because `()` is a subexpression, it captures the output of the command, but when running the external command `nu` thinks "oh, stdout is redirected, so there must be some other commands after me in the pipeline" and ignores the exit code :(

**Comment by gergelyk** [comment]
Good points, @abusch. I agree. Pipeline should start all the commands concurrently. In bash, how it works is that the first command that fails closes stdout. This is how subsequent command knows to exit (and close it's stdout too).

I think, what is important is the exit code that underlying shell should return. From what I observed:

1. In case of bash, by default, without `set -o pipefail`

    Shell returns the code returned by the last command in the pipeline.

2. In case of bash with `set -o pipefail`

    Shell returns the last non-zero code returned by the commands in the pipeline.

3. In case of nushell

    Shell returns the code returned by the last command in the pipeline. Just as bash does by default.

4. In case of elvish

    Shell returns 0 if all the commands returned 0, otherwise shell returns 2.
    Additionally, error message is printed to stderr. It reports all the non-zero codes returned by the commands in the pipeline.

Note: 1. and 2. is the same regardless or of `set -e`.

What I would consider for nushell it none of the behaviours above. My suggestion would be to return the _first_ non-zero code encountered in the pipeline. This gives some information about the root cause of the failure.

**Comment by lucio-ferrao** [comment]
I always start my shell scripts with 

```
#!/bin/bash

set -e
set -o pipefail
```


It would be great to have a similar option for nushell. I assumed it had a similar behavior until my scripts started failing with silent errors rather hard to catch...

**Comment by liff** [comment]
Regarding:

> Just wrap your calls to executables in `do -c { … }`

This behaves slightly differently from `(…)` wrt. standard error.

```nushell
> let x = (nu -c '(1..3) | each {|n| print -e $n; sleep 1sec}'; "x")
1
2
3
>
```

The standard error of the subexpression is attached to the terminal, but:

```nushell
> let x = do -c { nu -c '(1..3) | each {|n| print -e $n; sleep 1sec}'; "x" }
> 
```

`do` captures the standard error and only shows it in the error message if one occurs.

This can be relevant if the external command being run uses stderror to communicate progress information, for example.

**Comment by NfNitLoop** [comment]
Nu seems to have improved here a bit since I opened this issue.

Using 0.112.1, one of my original examples:
```
cat asdfasdf | split row " " | print; print done
```
now stops execution before outputting an empty table, and does not continue to print "done". ✅

Interestingly, `do -c` doesn't seem to give the nice `error make` style error messages anymore, and just gives:
```
> do -c { ex1 }
start
cat: asdfasdf: No such file or directory
# Also returns a non-zero exit code
```
... but it still stops on a non-zero error code, so that's fine by me. ✅

And sub-shells(?) no longer cause exit codes to be ignored! 🎉 

```
> (ex1)
start
cat: asdfasdf: No such file or directory
# Also returns a non-zero exit code
```

I'd say my most pressing concerns have been addressed -- execution stops when external tools return non-zero exit codes.

But I'll leave this issue open because it would be really nice, to get the `error make`-style errors pointing at *which* command stopped the flow of execution.

**Comment by Bahex** [comment]
@NfNitLoop Setting `$env.config.display_errors.exit_code = true` will give you the fancy errors for failing external commands too.
